### PR TITLE
feat: add adaptive sampling to python sdk

### PR DESCRIPTION
**Pull request by jy-tan**
### Summary

Add the Python half of the adaptive sampling rollout so the SDK can reduce recording pressure based on local queue, exporter, and memory signals without sacrificing pre-app-start capture or backward compatibility with existing sampling configuration.

### Changes

- Parse the new nested `recording.sampling` config while keeping legacy `sampling_rate` behavior backwards compatible
- Add an adaptive sampling controller and health loop driven by queue fill, dropped spans, exporter failures and circuit state, and memory pressure
- Suppress child span creation when inbound requests are skipped by admission decisions while continuing to always record pre-app-start traffic
- Standardize the canonical sampling env var on `TUSK_RECORDING_SAMPLING_RATE` while continuing to accept `TUSK_SAMPLING_RATE` as a backward-compatible alias
- Retain compatibility for older `_determine_sampling_rate` call sites and expand unit coverage for adaptive sampling, config loading, mode handling, env-var precedence, and no-record span suppression
- Add benchmark log analysis tooling plus related ignore rules to support tuning sampling-rate performance tradeoffs

### CODE REVIEW

File: drift/core/drift_sdk.py [hunk 1]
```diff
@@ -318,38 +338,171 @@ def initialize(
...
+        if config_sampling and config_sampling.mode in {"fixed", "adaptive"}:
+            mode = "adaptive" if config_sampling.mode == "adaptive" else "fixed"
+        elif config_sampling and config_sampling.mode:
+            logger.warning(
+                "Invalid sampling mode from config file: %s. Must be 'fixed' or 'adaptive'. Ignoring.",
+                config_sampling.mode,
+            )
+
+        base_rate = 1.0
         if init_param is not None:
```
Comment (ghost) [inline drift/core/drift_sdk.py:354]: P2: Behavioral regression: invalid `init_param` no longer falls back to env var or config file. In the old `_determine_sampling_rate`, when `init_param` was not None but invalid (e.g., 2.0), validation returned None and the code fell through to check `TUSK_SAMPLING_RATE` env var and then the config file. In the new code, the `if init_param is not None: ... else: ...` structure means an invalid init_param skips the entire `else` branch (env var, config_sampling.base_rate, recording_config.sampling_rate), silently defaulting `base_rate` to 1.0. A user passing an out-of-range sampling_rate to `initialize()` would get 100% sampling instead of the configured fallback.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>356</line_numbers>

<comment>Behavioral regression: invalid `init_param` no longer falls back to env var or config file. In the old `_determine_sampling_rate`, when `init_param` was not None but invalid (e.g., 2.0), validation returned None and the code fell through to check `TUSK_SAMPLING_RATE` env var and then the config file. In the new code, the `if init_param is not None: ... else: ...` structure means an invalid init_param skips the entire `else` branch (env var, config_sampling.base_rate, recording_config.sampling_rate), silently defaulting `base_rate` to 1.0. A user passing an out-of-range sampling_rate to `initialize()` would get 100% sampling instead of the configured fallback.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>356</line_numbers>

<comment>Behavioral regression: invalid `init_param` no longer falls back to env var or config file. In the old `_determine_sampling_rate`, when `init_param` was not None but invalid (e.g., 2.0), validation returned None and the code fell through to check `TUSK_SAMPLING_RATE` env var and then the config file. In the new code, the `if init_param is not None: ... else: ...` structure means an invalid init_param skips the entire `else` branch (env var, config_sampling.base_rate, recording_config.sampling_rate), silently defaulting `base_rate` to 1.0. A user passing an out-of-range sampling_rate to `initialize()` would get 100% sampling instead of the configured fallback.</comment>
```

</details>

File: drift/core/drift_sdk.py [hunk 2]
```diff
@@ -318,38 +338,171 @@ def initialize(
...
+
+        self._adaptive_sampling_thread = threading.Thread(
+            target=self._adaptive_sampling_loop,
+            daemon=True,
+            name="drift-adaptive-sampling",
+        )
+        self._adaptive_sampling_thread.start()
+        self._update_adaptive_sampling_health()
+
+    def _adaptive_sampling_loop(self) -> None:
```
Comment (ghost) [inline drift/core/drift_sdk.py:433]: P2: The adaptive sampling background thread has no exception handling. If `_update_adaptive_sampling_health` raises any exception (e.g., an attribute access on a span processor or exporter that's been partially shut down, or any transient error), the exception propagates out of the `while` loop and the thread terminates permanently. Adaptive sampling health monitoring would silently stop for the rest of the process lifetime, leaving the controller frozen at its last state.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>428</line_numbers>

<comment>The adaptive sampling background thread has no exception handling. If `_update_adaptive_sampling_health` raises any exception (e.g., an attribute access on a span processor or exporter that's been partially shut down, or any transient error), the exception propagates out of the `while` loop and the thread terminates permanently. Adaptive sampling health monitoring would silently stop for the rest of the process lifetime, leaving the controller frozen at its last state.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>428</line_numbers>

<comment>The adaptive sampling background thread has no exception handling. If `_update_adaptive_sampling_health` raises any exception (e.g., an attribute access on a span processor or exporter that's been partially shut down, or any transient error), the exception propagates out of the `while` loop and the thread terminates permanently. Adaptive sampling health monitoring would silently stop for the rest of the process lifetime, leaving the controller frozen at its last state.</comment>
```

</details>

File: drift/instrumentation/wsgi/handler.py [hunk 3]
```diff
@@ -225,7 +226,8 @@ def _create_and_handle_request(
         )
         if not should_record:
             logger.debug(f"[WSGI] Skipping request ({skip_reason}), path={path}")
-            return original_wsgi_app(app, environ, start_response)
+            with suppress_recording():
```
Comment (ghost) [inline drift/instrumentation/wsgi/handler.py:253]: P3: WSGI `suppress_recording()` context manager exits before the response iterator is consumed. WSGI apps return an `Iterable[bytes]` that the server consumes lazily. The `with suppress_recording(): return original_wsgi_app(...)` pattern resets the ContextVar when the `with` block exits (at `return`), but the returned iterable is consumed by the WSGI server after the function returns. Any span-creating code that runs during response body iteration (e.g., streaming middleware, instrumented generators) will NOT see the suppression flag. Django and FastAPI don't have this issue because they fully process the response before returning.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/instrumentation/wsgi/handler.py</file_path>
<line_numbers>229</line_numbers>

<comment>WSGI `suppress_recording()` context manager exits before the response iterator is consumed. WSGI apps return an `Iterable[bytes]` that the server consumes lazily. The `with suppress_recording(): return original_wsgi_app(...)` pattern resets the ContextVar when the `with` block exits (at `return`), but the returned iterable is consumed by the WSGI server after the function returns. Any span-creating code that runs during response body iteration (e.g., streaming middleware, instrumented generators) will NOT see the suppression flag. Django and FastAPI don't have this issue because they fully process the response before returning.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/instrumentation/wsgi/handler.py</file_path>
<line_numbers>229</line_numbers>

<comment>WSGI `suppress_recording()` context manager exits before the response iterator is consumed. WSGI apps return an `Iterable[bytes]` that the server consumes lazily. The `with suppress_recording(): return original_wsgi_app(...)` pattern resets the ContextVar when the `with` block exits (at `return`), but the returned iterable is consumed by the WSGI server after the function returns. Any span-creating code that runs during response body iteration (e.g., streaming middleware, instrumented generators) will NOT see the suppression flag. Django and FastAPI don't have this issue because they fully process the response before returning.</comment>
```

</details>

File: drift/core/drift_sdk.py [hunk 4]
```diff
@@ -318,38 +338,171 @@ def initialize(
...
+            return cgroup_current / self._effective_memory_limit_bytes
+
+        cgroup_v1_current = self._read_numeric_control_file("/sys/fs/cgroup/memory/memory.usage_in_bytes")
+        if cgroup_v1_current is not None:
+            return cgroup_v1_current / self._effective_memory_limit_bytes
+
+        try:
+            import resource
+
+            rss = resource.getrusage(resource.RUSAGE_SELF).ru_maxrss
```
Comment (ghost) [inline drift/core/drift_sdk.py]: P3: The fallback memory pressure calculation uses `resource.getrusage(resource.RUSAGE_SELF).ru_maxrss`, which returns the **peak** (high-water mark) resident set size, not the **current** RSS. This means memory_pressure_ratio will reflect the highest memory usage ever seen by the process, not the current usage. This permanently overestimates memory pressure and could trigger unnecessary sampling rate reductions or critical pauses that never recover, since peak RSS only increases over time.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>490</line_numbers>

<comment>The fallback memory pressure calculation uses `resource.getrusage(resource.RUSAGE_SELF).ru_maxrss`, which returns the **peak** (high-water mark) resident set size, not the **current** RSS. This means memory_pressure_ratio will reflect the highest memory usage ever seen by the process, not the current usage. This permanently overestimates memory pressure and could trigger unnecessary sampling rate reductions or critical pauses that never recover, since peak RSS only increases over time.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>490</line_numbers>

<comment>The fallback memory pressure calculation uses `resource.getrusage(resource.RUSAGE_SELF).ru_maxrss`, which returns the **peak** (high-water mark) resident set size, not the **current** RSS. This means memory_pressure_ratio will reflect the highest memory usage ever seen by the process, not the current usage. This permanently overestimates memory pressure and could trigger unnecessary sampling rate reductions or critical pauses that never recover, since peak RSS only increases over time.</comment>
```

</details>

File: drift/core/drift_sdk.py [hunk 5]
```diff
@@ -801,6 +954,34 @@ async def send_unpatched_dependency_alert(
         except Exception as e:
             logger.debug(f"Failed to send unpatched dependency alert: {e}")
 
+    def should_record_root_request(self, *, is_pre_app_start: bool) -> RootSamplingDecision:
+        if self._adaptive_sampling_controller is not None:
```
Comment (ghost) [inline drift/core/drift_sdk.py:1003]: P4: The `should_record_root_request` method and `_update_adaptive_sampling_health` operate on the `AdaptiveSamplingController` from different threads (request threads and the background health thread) without synchronization. The `update()` method modifies `_admission_multiplier`, `_state`, and `_paused_until_s` in a non-atomic sequence, while `get_decision()` and `get_effective_sampling_rate()` read multiple fields. A request thread could observe `_state='critical_pause'` but `_admission_multiplier=1.0` (or vice versa) if it reads between writes in `update()`. Under CPython's GIL individual attribute access is atomic, but reading a consistent snapshot of multiple fields is not guaranteed.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>958</line_numbers>

<comment>The `should_record_root_request` method and `_update_adaptive_sampling_health` operate on the `AdaptiveSamplingController` from different threads (request threads and the background health thread) without synchronization. The `update()` method modifies `_admission_multiplier`, `_state`, and `_paused_until_s` in a non-atomic sequence, while `get_decision()` and `get_effective_sampling_rate()` read multiple fields. A request thread could observe `_state='critical_pause'` but `_admission_multiplier=1.0` (or vice versa) if it reads between writes in `update()`. Under CPython's GIL individual attribute access is atomic, but reading a consistent snapshot of multiple fields is not guaranteed.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>958</line_numbers>

<comment>The `should_record_root_request` method and `_update_adaptive_sampling_health` operate on the `AdaptiveSamplingController` from different threads (request threads and the background health thread) without synchronization. The `update()` method modifies `_admission_multiplier`, `_state`, and `_paused_until_s` in a non-atomic sequence, while `get_decision()` and `get_effective_sampling_rate()` read multiple fields. A request thread could observe `_state='critical_pause'` but `_admission_multiplier=1.0` (or vice versa) if it reads between writes in `update()`. Under CPython's GIL individual attribute access is atomic, but reading a consistent snapshot of multiple fields is not guaranteed.</comment>
```

</details>

File: drift/core/mode_utils.py [hunk 6]
```diff
@@ -187,11 +187,10 @@ def should_record_inbound_http_request(
...
     if not is_pre_app_start:
         from .drift_sdk import TuskDrift
-        from .sampling import should_sample
 
         sdk = TuskDrift.get_instance()
-        sampling_rate = sdk.get_sampling_rate()
-        if not should_sample(sampling_rate, is_app_ready=True):
+        decision = sdk.should_record_root_request(is_pre_app_start=is_pre_app_start)
+        if not decision.should_record:
             return False, "not_sampled"
```
Comment (ghost) [inline drift/core/mode_utils.py]: P4: `should_record_inbound_http_request` always returns the hardcoded string `"not_sampled"` when the adaptive sampling controller decides to skip a request, discarding the actual `decision.reason` which could be `"critical_pause"` or `"load_shed"`. All three frameworks (Django, FastAPI, WSGI) log this reason in their debug output. When the system is in a critical_pause state (e.g., memory pressure or circuit breaker open), operators will see 'not_sampled' in logs instead of the more informative 'critical_pause', making it harder to diagnose adaptive sampling behavior.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/mode_utils.py</file_path>
<line_numbers>194</line_numbers>

<comment>`should_record_inbound_http_request` always returns the hardcoded string `"not_sampled"` when the adaptive sampling controller decides to skip a request, discarding the actual `decision.reason` which could be `"critical_pause"` or `"load_shed"`. All three frameworks (Django, FastAPI, WSGI) log this reason in their debug output. When the system is in a critical_pause state (e.g., memory pressure or circuit breaker open), operators will see 'not_sampled' in logs instead of the more informative 'critical_pause', making it harder to diagnose adaptive sampling behavior.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/mode_utils.py</file_path>
<line_numbers>194</line_numbers>

<comment>`should_record_inbound_http_request` always returns the hardcoded string `"not_sampled"` when the adaptive sampling controller decides to skip a request, discarding the actual `decision.reason` which could be `"critical_pause"` or `"load_shed"`. All three frameworks (Django, FastAPI, WSGI) log this reason in their debug output. When the system is in a critical_pause state (e.g., memory pressure or circuit breaker open), operators will see 'not_sampled' in logs instead of the more informative 'critical_pause', making it harder to diagnose adaptive sampling behavior.</comment>
```

</details>

File: drift/core/adaptive_sampling.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+    def update(self, snapshot: AdaptiveSamplingHealthSnapshot) -> None:
+        with self._lock:
+            if self._config.mode != "adaptive":
+                self._state = "fixed"
+                self._admission_multiplier = 1.0
+                return
+
+            now_s = self._now_fn()
+            elapsed_s = 2.0 if self._last_updated_at_s == 0 else max(0.001, now_s - self._last_updated_at_s)
```
Comment (ghost) [inline drift/core/adaptive_sampling.py]: P4: `AdaptiveSamplingController` uses `0` as a sentinel value for `_last_updated_at_s` to detect the first `update()` call: `elapsed_s = 2.0 if self._last_updated_at_s == 0 else max(0.001, now_s - self._last_updated_at_s)`. However, after the first `update()` sets `self._last_updated_at_s = now_s`, if `now_fn()` returned `0.0` (as it does in tests with `now_fn=lambda: 0.0`), the sentinel check `0.0 == 0` is `True` in Python. This causes every subsequent call to also use the default `elapsed_s = 2.0` instead of computing the real elapsed time, producing incorrect EWMA decay rates for queue fill, drop, and failure signals.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/adaptive_sampling.py</file_path>
<line_numbers>102</line_numbers>

<comment>`AdaptiveSamplingController` uses `0` as a sentinel value for `_last_updated_at_s` to detect the first `update()` call: `elapsed_s = 2.0 if self._last_updated_at_s == 0 else max(0.001, now_s - self._last_updated_at_s)`. However, after the first `update()` sets `self._last_updated_at_s = now_s`, if `now_fn()` returned `0.0` (as it does in tests with `now_fn=lambda: 0.0`), the sentinel check `0.0 == 0` is `True` in Python. This causes every subsequent call to also use the default `elapsed_s = 2.0` instead of computing the real elapsed time, producing incorrect EWMA decay rates for queue fill, drop, and failure signals.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/adaptive_sampling.py</file_path>
<line_numbers>102</line_numbers>

<comment>`AdaptiveSamplingController` uses `0` as a sentinel value for `_last_updated_at_s` to detect the first `update()` call: `elapsed_s = 2.0 if self._last_updated_at_s == 0 else max(0.001, now_s - self._last_updated_at_s)`. However, after the first `update()` sets `self._last_updated_at_s = now_s`, if `now_fn()` returned `0.0` (as it does in tests with `now_fn=lambda: 0.0`), the sentinel check `0.0 == 0` is `True` in Python. This causes every subsequent call to also use the default `elapsed_s = 2.0` instead of computing the real elapsed time, producing incorrect EWMA decay rates for queue fill, drop, and failure signals.</comment>
```

</details>

File: drift/core/adaptive_sampling.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        snapshot: AdaptiveSamplingHealthSnapshot,
+    ) -> None:
+        if previous_state == self._state and abs(previous_multiplier - self._admission_multiplier) < 0.05:
+            return
+
+        logger.info(
+            "Adaptive sampling updated (state=%s, multiplier=%.2f, effective_rate=%.4f, pressure=%.2f, queue_fill=%s, memory_pressure_ratio=%s, export_circuit_open=%s).",
+            self._state,
+            self._admission_multiplier,
+            self.get_effective_sampling_rate(),
```
Comment (ghost) [inline drift/core/adaptive_sampling.py:249]: P3: `_log_transition` calls `self.get_effective_sampling_rate()` while `update()` already holds `self._lock`. `get_effective_sampling_rate()` calls `self._now_fn()` a second time, which can return a different value than the `now_s` captured at the top of `update()`. In the critical_pause path, `update()` sets `_paused_until_s = now_s + 15`, then `_log_transition` calls `get_effective_sampling_rate()` which re-evaluates `self._now_fn() < self._paused_until_s`. If `_now_fn()` advances (mocked time in tests, or measurable elapsed time in production under load), the logged effective_rate will be inconsistent with the state transition that was just computed.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/adaptive_sampling.py</file_path>
<line_numbers>249</line_numbers>

<comment>`_log_transition` calls `self.get_effective_sampling_rate()` while `update()` already holds `self._lock`. `get_effective_sampling_rate()` calls `self._now_fn()` a second time, which can return a different value than the `now_s` captured at the top of `update()`. In the critical_pause path, `update()` sets `_paused_until_s = now_s + 15`, then `_log_transition` calls `get_effective_sampling_rate()` which re-evaluates `self._now_fn() < self._paused_until_s`. If `_now_fn()` advances (mocked time in tests, or measurable elapsed time in production under load), the logged effective_rate will be inconsistent with the state transition that was just computed.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/adaptive_sampling.py</file_path>
<line_numbers>249</line_numbers>

<comment>`_log_transition` calls `self.get_effective_sampling_rate()` while `update()` already holds `self._lock`. `get_effective_sampling_rate()` calls `self._now_fn()` a second time, which can return a different value than the `now_s` captured at the top of `update()`. In the critical_pause path, `update()` sets `_paused_until_s = now_s + 15`, then `_log_transition` calls `get_effective_sampling_rate()` which re-evaluates `self._now_fn() < self._paused_until_s`. If `_now_fn()` advances (mocked time in tests, or measurable elapsed time in production under load), the logged effective_rate will be inconsistent with the state transition that was just computed.</comment>
```

</details>

File: drift/core/drift_sdk.py [hunk 9]
```diff
@@ -318,38 +336,222 @@ def initialize(
...
-            validated = validate_sampling_rate(config_rate, "config file")
+        if base_rate is None:
+            for env_key in ("TUSK_RECORDING_SAMPLING_RATE", "TUSK_SAMPLING_RATE"):
+                env_rate = os.environ.get(env_key)
+                if env_rate is None:
+                    continue
+
+                try:
+                    parsed = float(env_rate)
+                    validated = validate_sampling_rate(parsed, f"{env_key} env var")
```
Comment (ghost) [inline drift/core/drift_sdk.py:368]: P4: When `TUSK_RECORDING_SAMPLING_RATE` is set to a syntactically valid but out-of-range value (e.g., `2.0`), `validate_sampling_rate` returns `None`, and the loop silently falls through to try `TUSK_SAMPLING_RATE`. The user gets no indication that the canonical env var was rejected in favor of the legacy alias. `validate_sampling_rate` does log a generic warning about the invalid rate, but it doesn't mention that a fallback occurred. A user setting both env vars with different values may be surprised by which rate is actually used.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>368</line_numbers>

<comment>When `TUSK_RECORDING_SAMPLING_RATE` is set to a syntactically valid but out-of-range value (e.g., `2.0`), `validate_sampling_rate` returns `None`, and the loop silently falls through to try `TUSK_SAMPLING_RATE`. The user gets no indication that the canonical env var was rejected in favor of the legacy alias. `validate_sampling_rate` does log a generic warning about the invalid rate, but it doesn't mention that a fallback occurred. A user setting both env vars with different values may be surprised by which rate is actually used.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>drift/core/drift_sdk.py</file_path>
<line_numbers>368</line_numbers>

<comment>When `TUSK_RECORDING_SAMPLING_RATE` is set to a syntactically valid but out-of-range value (e.g., `2.0`), `validate_sampling_rate` returns `None`, and the loop silently falls through to try `TUSK_SAMPLING_RATE`. The user gets no indication that the canonical env var was rejected in favor of the legacy alias. `validate_sampling_rate` does log a generic warning about the invalid rate, but it doesn't mention that a fallback occurred. A user setting both env vars with different values may be surprised by which rate is actually used.</comment>
```

</details>

File: tests/unit/test_adaptive_sampling.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+
+    now["value"] = 1.0
+    controller.update(AdaptiveSamplingHealthSnapshot(queue_fill_ratio=0.1, dropped_span_count=1))
+    paused_decision = controller.get_decision(is_pre_app_start=False)
+    assert paused_decision.state == "critical_pause"
+    assert paused_decision.should_record is False
+    assert paused_decision.reason == "critical_pause"
+
+
+def test_elapsed_time_uses_zero_timestamp_as_real_value():
```
Comment (ghost) [inline tests/unit/test_adaptive_sampling.py:48]: P4: Test name `test_elapsed_time_uses_zero_timestamp_as_real_value` claims to verify that `now_fn()` returning `0.0` is treated as a valid timestamp. However, the test doesn't exercise this edge case. The first `update()` uses the default `elapsed_s=2.0` (because `_last_updated_at_s is None`), and the second uses `elapsed_s=0.5` (0.5-0.0). The test only validates EWMA decay math, not the zero-timestamp boundary. To actually test the claim, both updates should use `now=0.0` to verify `elapsed_s=max(0.001, 0.0-0.0)=0.001`.

<details>
<summary>AI prompts</summary>

**Check prompt**
```text
Check if this issue is valid. If so, find the root cause and let me know. Don't update any code just yet.

<file_path>tests/unit/test_adaptive_sampling.py</file_path>
<line_numbers>48</line_numbers>

<comment>Test name `test_elapsed_time_uses_zero_timestamp_as_real_value` claims to verify that `now_fn()` returning `0.0` is treated as a valid timestamp. However, the test doesn't exercise this edge case. The first `update()` uses the default `elapsed_s=2.0` (because `_last_updated_at_s is None`), and the second uses `elapsed_s=0.5` (0.5-0.0). The test only validates EWMA decay math, not the zero-timestamp boundary. To actually test the claim, both updates should use `now=0.0` to verify `elapsed_s=max(0.001, 0.0-0.0)=0.001`.</comment>
```

**Check + fix prompt**
```text
Check if this issue is valid. If so, find the root cause and fix it.

<file_path>tests/unit/test_adaptive_sampling.py</file_path>
<line_numbers>48</line_numbers>

<comment>Test name `test_elapsed_time_uses_zero_timestamp_as_real_value` claims to verify that `now_fn()` returning `0.0` is treated as a valid timestamp. However, the test doesn't exercise this edge case. The first `update()` uses the default `elapsed_s=2.0` (because `_last_updated_at_s is None`), and the second uses `elapsed_s=0.5` (0.5-0.0). The test only validates EWMA decay math, not the zero-timestamp boundary. To actually test the claim, both updates should use `now=0.0` to verify `elapsed_s=max(0.001, 0.0-0.0)=0.001`.</comment>
```

</details>

**Review by sohil-kshirsagar — APPROVED** [review]
